src/admin_ws.rs

```diff
@@ -0,0 +1,203 @@
+use holochain_client::{AdminWebsocket, ConductorApiError, ConductorApiResult};
+use std::{future::Future, sync::Arc};
+use tokio::sync::RwLock;
+use url2::Url2;
+
+use crate::{HcHttpGatewayError, HcHttpGatewayResult};
+
+const ADMIN_WS_CONNECTION_MAX_RETRIES: usize = 1;
+
+/// A wrapper around AdminWebsocket that automatically handles reconnection
+/// when the connection is lost due to network issues or other failures.
+#[derive(Debug, Clone)]
+pub struct HcHttpGwAdminWebsocket {
+    /// The WebSocket URL to connect to
+    url: Url2,
+    /// The handle to the AdminWebsocket connection
+    connection_handle: Arc<RwLock<Option<AdminWebsocket>>>,
+    /// Current retry attempt counter
+    current_retries: usize,
+}
+
+impl HcHttpGwAdminWebsocket {
+    /// Creates a new HcHttpGwAdminWebsocket and attempts to connect immediately
+    ///
+    /// This will make multiple attempts according to the `max_retries`
+    /// setting, with exponential backoff between retries.
+    pub async fn connect(url: &Url2) -> HcHttpGatewayResult<Self> {
+        let mut instance = Self {
+            url: url.clone(),
+            connection_handle: Arc::new(RwLock::new(None)),
+            current_retries: 0,
+        };
+
+        instance.attempt_connection().await?;
+        Ok(instance)
+    }
+
+    /// Checks if there is an active connection
+    async fn is_connected(&self) -> HcHttpGatewayResult<bool> {
+        let connection = self.connection_handle.read().await;
+        Ok(connection.is_some())
+    }
+
+    /// Gets reconnection retry count
+    pub fn get_reconnection_retries(&self) -> usize {
+        self.current_retries
+    }
+
+    /// Ensures that a connection is established before proceeding
+    async fn ensure_connected(&mut self) -> HcHttpGatewayResult<()> {
+        if self.is_connected().await? {
+            return Ok(());
+        }
+
+        self.attempt_connection().await
+    }
+
+    /// Attempts to connect to the AdminWebsocket
+    ///
+    /// This will make multiple attempts according to the `max_retries`
+    /// setting, with exponential backoff between retries.
+    async fn attempt_connection(&mut self) -> HcHttpGatewayResult<()> {
+        self.current_retries = 0;
+
+        while self.current_retries < ADMIN_WS_CONNECTION_MAX_RETRIES {
```

Comment:
```suggestion
        while self.current_retries <= ADMIN_WS_CONNECTION_MAX_RETRIES {
```

I think this is the intention